src/engine/sparqlExpressions/StringExpressions.cpp

```diff
@@ -209,37 +220,46 @@ class SubstrImpl {
     if (startInt < 0) {
       lengthInt += startInt;
     }
-
     const auto& str = asStringViewUnsafe(s.value().getContent());
+    std::size_t utf8len = utf8Length(str);
     // Clamp the number such that it is in `[0, str.size()]`. That way we end up
-    // with valid arguments for the `getUTF8Substring` method below for both
+    // with valid arguments for the `setSubstr` method below for both
     // starting position and length since all the other corner cases have been
     // dealt with above.
-    auto clamp = [sz = str.size()](int64_t n) -> std::size_t {
+    auto clamp = [utf8len](int64_t n) -> std::size_t {
       if (n < 0) {
         return 0;
       }
-      if (static_cast<size_t>(n) > sz) {
-        return sz;
+      if (static_cast<size_t>(n) > utf8len) {
+        return utf8len;
       }
       return static_cast<size_t>(n);
     };
 
-    s.value().getLiteral().setSubstr(clamp(startInt), clamp(lengthInt));
-    return s.value();
+    startInt = clamp(startInt);
+    lengthInt = clamp(lengthInt);
```

Comment:
I think you have to use a clamped version of `start + length` , ottherwise this doesn't quite work.